pkg/apis/pingcap/v1alpha1/tidbautoscaler_types.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"k8s.io/api/autoscaling/v2beta2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// +k8s:openapi-gen=true
+// TidbClusterAutoScaler is the control script's spec
+type TidbClusterAutoScaler struct {
+	metav1.TypeMeta `json:",inline"`
+	// +k8s:openapi-gen=false
+	metav1.ObjectMeta `json:"metadata"`
+
+	// Spec describes the state of the TidbClusterAutoScaler
+	Spec TidbClusterAutoScalerSpec `json:"spec"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// +k8s:openapi-gen=true
+// TidbClusterAutoScalerList is TidbClusterAutoScaler list
+type TidbClusterAutoScalerList struct {
+	metav1.TypeMeta `json:",inline"`
+	// +k8s:openapi-gen=false
+	metav1.ListMeta `json:"metadata"`
+
+	Items []TidbClusterAutoScaler `json:"items"`
+}
+
+// +k8s:openapi-gen=true
+// TidbAutoScalerSpec describes the
+type TidbClusterAutoScalerSpec struct {
+	// TidbClusterRef
+	Cluster TidbClusterRef `json:"cluster"`
+
+	// We used prometheus to fetch the metrics resources until the pd could provide it.
+	// MetricsUrl represents the url to fetch the metrics info
+	// +optional
+	MetricsUrl *string `json:"metricsUrl,omitempty"`
+
+	// Tikv represents the auto-scaling spec for tikv
+	Tikv TikvAutoScalerSpec `json:"tikv"`
+
+	// Tidb represents the auto-scaling spec for tidb
+	Tidb TidbAutoScalerSpec `json:"tidb"`
+}
+
+// +k8s:openapi-gen=true
+// TikvAutoScalerSpec describes the spec for tikv auto-scaling
+type TikvAutoScalerSpec struct {
+	// maxReplicas is the upper limit for the number of replicas to which the autoscaler can scale up.
+	// It cannot be less that minReplicas.
+	MaxReplicas int32 `json:"maxReplicas" protobuf:"varint,3,opt,name=maxReplicas"`
+
+	// minReplicas is the lower limit for the number of replicas to which the autoscaler
+	// can scale down.  It defaults to 1 pod. Scaling is active as long as at least one metric value is
+	// available.
+	// +optional
+	MinReplicas *int32 `json:"minReplicas,omitempty" protobuf:"varint,2,opt,name=minReplicas"`
+
+	// IntervalDurationSeconds represents the duration seconds between each auto-scaling interval
```

Comment:
There should be another duration to mitigate the scale jitter.

Reply:
updated.

Reply:
Which paremeter is updated?